cmd/ingester/app/consumer/consumer.go

```diff
@@ -44,50 +60,149 @@ type Consumer struct {
 	internalConsumer consumer.Consumer
 	processorFactory ProcessorFactory
 
-	deadlockDetector deadlockDetector
+	deadlockDetector              deadlockDetector
+	deadlockNotifyCh              chan struct{}
+	deadlockNotifyOnce            sync.Once
+	globalDeadlockDetectorEnabled bool // defailt is enabled
 
-	partitionIDToState  map[int32]*consumerState
-	partitionMapLock    sync.Mutex
-	partitionsHeld      int64
+	partitionsHeld      atomic.Int64
 	partitionsHeldGauge metrics.Gauge
 
+	waitReady     bool // defailt is not wait
+	consumerReady *consumerReady
+
 	doneWg sync.WaitGroup
+
+	topic  string
+	cancel context.CancelFunc
 }
 
-type consumerState struct {
-	partitionConsumer sc.PartitionConsumer
+type consumerReady struct {
+	readyCh   chan struct{}
+	closeOnce sync.Once
+}
+
+func (c *consumerReady) waitReady() {
+	<-c.readyCh
+}
+
+func (c *consumerReady) markReady() {
+	c.close()
+}
+
+func (c *consumerReady) close() {
+	c.closeOnce.Do(func() {
+		close(c.readyCh)
+	})
 }
 
 // New is a constructor for a Consumer
-func New(params Params) (*Consumer, error) {
+func New(params Params, opts ...Option) (*Consumer, error) {
 	deadlockDetector := newDeadlockDetector(params.MetricsFactory, params.Logger, params.DeadlockCheckInterval)
-	return &Consumer{
-		metricsFactory:      params.MetricsFactory,
-		logger:              params.Logger,
-		internalConsumer:    params.InternalConsumer,
-		processorFactory:    params.ProcessorFactory,
-		deadlockDetector:    deadlockDetector,
-		partitionIDToState:  make(map[int32]*consumerState),
-		partitionsHeldGauge: partitionsHeldGauge(params.MetricsFactory),
-	}, nil
-}
-
-// Start begins consuming messages in a go routine
+	c := &Consumer{
+		metricsFactory:                params.MetricsFactory,
+		logger:                        params.Logger,
+		internalConsumer:              params.InternalConsumer,
+		processorFactory:              params.ProcessorFactory,
+		deadlockDetector:              deadlockDetector,
+		deadlockNotifyCh:              make(chan struct{}, 1),
+		globalDeadlockDetectorEnabled: true,
+		partitionsHeldGauge:           partitionsHeldGauge(params.MetricsFactory),
+		consumerReady: &consumerReady{
+			readyCh: make(chan struct{}, 1),
+		},
+		waitReady: false,
+		topic:     params.ProcessorFactory.topic,
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c, nil
+}
+
+// doStart begins consuming messages in a go routine
 func (c *Consumer) Start() {
-	c.deadlockDetector.start()
+	// Start begins consuming and wait is running and disable global deadlock detector,
```

Comment:
Did you mean to place this comment above the `Start()` func? If so, can you please reword it as I'm having a little trouble parsing this comment block. e.g. i would replace `specially partition` -> `partition-specific`.